dissect: single-filesystem verity images with external hashdevice

bluca opened the pull request:
dm-verity support in dissect-image at the moment is restricted to GPT
volumes.
If the image a single-filesystem type without a partition table (eg: squashfs)
and a roothash/hashdevice are passed, set the verity flag and mark as
read-only.

bluca commented:
Wanted to add a systemd-dissect --mount to the unit test, but unfortunately in the test environment it seems opening verity devices is completely broken, and hangs forever. This is even with just /sbin/veritysetup open foo bar - nothing useful in the logs, it just hangs.

bluca commented:
@mrc0mmand does the CentOS 7 CI image lack veritysetup/mksquashfs by any chance? It seems they are not being copied, and the setup step fails - this unit test infra really doesn't like me...

bluca commented:
I'll see if I can repro locally and then reopen later

poettering commented:
Hmm, not sure I grok what the patch is supposed to do? the dissect logic is supposed to set up verity for you (i.e. , dissected_image_decrypt() does that) and for that discovers both the fs itself and the verity data as two distinct GPT partitions in them. But if you have a raw file system image without GPT envelope, then where will it find the verity data?

to support verity in this mode properly, the dissected_image_decrypt() function would need to take an additional argument I guess pointing to a raw file with the verity hash data in it. And this would then have to be exposed in systemd-mount, so that you invoke it like this:

```
systemd-mount --roothash=81fde859ab844e389efc38232a667bf036f62c7a9cd24613b9ea0b22be62c91a --verity=theveritydata.raw thefs.raw
```

bluca commented:
> Hmm, not sure I grok what the patch is supposed to do? the dissect logic is supposed to set up verity for you (i.e. , dissected_image_decrypt() does that) and for that discovers both the fs itself and the verity data as two distinct GPT partitions in them. But if you have a raw file system image without GPT envelope, then where will it find the verity data?
> 
> to support verity in this mode properly, the dissected_image_decrypt() function would need to take an additional argument I guess pointing to a raw file with the verity hash data in it. And this would then have to be exposed in systemd-mount, so that you invoke it like this:
> 
> ```
> systemd-mount --roothash=81fde859ab844e389efc38232a667bf036f62c7a9cd24613b9ea0b22be62c91a --verity=theveritydata.raw thefs.raw /foo
> ```

It would probably help if I pushed the whole thing! Epic fail at git add -p :-)

I added support for an external hash device as well, which is necessary tipically with a squashfs volume.

bluca commented:
Also changed the test to skip if it cannot install all the modules/tools required, which hopefully will help avoid false negatives on CentOS 7.

bluca commented on src/shared/dissect-image.c:
```diff
@@ -1409,56 +1414,101 @@ int decrypted_image_relinquish(DecryptedImage *d) {
-        if (l < sizeof(sd_id128_t))
-                return -EINVAL;
+        if (hashdevice) {
+                r = getxattr_malloc(image, "user.verity.hashdevice", hashdevice, true);
```
I picked .hashdevice as the extension to match util-linux's command line option names, to have some cross-tool consistency for handling the same objects. But if there's a preference for a different naming scheme I'm ok with changing to something else.

poettering replied:
hmm, i wonder if this should be an xattr at all... i mean, you are looking for a file here, and hence might as well look for that right-away. i.e. the indirection via an xattr seems unnecessary. This is different for the roothash which itself isn't a path but metadata directly associated with the image where an xattr makes a lot more sense

bluca replied:
Sure, was just following the same pattern. Will remove.

poettering commented:
btw, what's the rationale for using raw disk files without GPT envelope? wouldn't it be a ton nicer and simpler and easier to just pack them into the same image file with a GPT envelope?

bluca commented:
> btw, what's the rationale for using raw disk files without GPT envelope? wouldn't it be a ton nicer and simpler and easier to just pack them into the same image file with a GPT envelope?

Not 100% sure, as the decision to use raw squashfs was done at the very beginning, a bit before I joined this project - my best guess is that it's already hard enough to get applications cross-compiled correctly from Windows (yes sadly that was a hard requirement) and delivered as squashfs + verity (+ selinux and a bunch of other funny things), adding GPT on top of that probably seemed not worth the hassle. Also the primary tool used during development was veritysetup, which  IIRC doesn't want a partition table, just the device - this probably helped drive the design too.

mrc0mmand commented:
The CentOS CI fails (Arch in KVM is going to fail as well) should be fixed once I merge https://github.com/systemd/systemd-centos-ci/pull/271, which should happen in about half an hour (just want to make sure the new Arch kernel doesn't break anything). I'll reschedule the failing jobs once it's done.

bluca commented:
> > The CentOS CI fails (Arch in KVM is going to fail as well) should be fixed once I merge [systemd/systemd-centos-ci#271](https://github.com/systemd/systemd-centos-ci/pull/271), which should happen in about half an hour (just want to make sure the new Arch kernel doesn't break anything). I'll reschedule the failing jobs once it's done.
> 
> Thanks!
> 
> Is the `( setup ) || exit 0` I added in the setup script not the right way to skip if the tools are missing? Didn't seem to work

Actually it seems to work fine, if I add an inst_binary which doesn't exist. The setup exits and the test is skipped.

The log from the CI said a cached image was reused - could it be stored from the previous run before I added the skip?

mrc0mmand commented:
> 
> Is the `( setup ) || exit 0` I added in the setup script not the right way to skip if the tools are missing? Didn't seem to work

Nope, since the test is executed in phases, like `make -C test/TEST-XXX clean setup run`. So in this case the `setup` phase calls `test_create_image()` and succeeds, and is followed by the `run` phase, which fails because of missing tools.

If one would really want to skip a test based on existence of a tool, the check would have to be outside of any function, so each phase goes through it, i.e. like TEST-06-SELINUX does it:

https://github.com/systemd/systemd/blob/112bed84bffdc60db8bae2aa27b60f35d49fe7f2/test/TEST-06-SELINUX/test.sh#L12-L13

But in this case installing the necessary dependencies is the best approach, since they're pretty common.

mrc0mmand commented:
>  `( setup ) || exit 0`

Also, this expression has a pretty nasty side-effect when combined with `set -e`:
>  If  a  compound  command or shell function executes in a context where -e is being ignored, none of the commands executed within the compound command or function body will be affected by the -e setting, even if -e is set and  a  command returns a failure status.  If a compound command or shell function sets -e while executing in a context where -e is ignored, that setting will not have any effect until the compound command or the command containing  the  function call completes.

(quote from `bash(1) man page for `set` command). 

For this reason we dropped it in cc469c3dfc398210f38f819d367e68646c71d8da as it could mask (and it did mask) some setup errors.

So, please, drop it as well, once you get to it :-)

bluca commented:
> > `( setup ) || exit 0`
> 
> Also, this expression has a pretty nasty side-effect when combined with `set -e`:
> 
> > If  a  compound  command or shell function executes in a context where -e is being ignored, none of the commands executed within the compound command or function body will be affected by the -e setting, even if -e is set and  a  command returns a failure status.  If a compound command or shell function sets -e while executing in a context where -e is ignored, that setting will not have any effect until the compound command or the command containing  the  function call completes.
> 
> (quote from `bash(1)` man page for `set` command).
> 
> For this reason we dropped it in [cc469c3](https://github.com/systemd/systemd/commit/cc469c3dfc398210f38f819d367e68646c71d8da) as it could mask (and it did mask) some setup errors.
> 
> So, please, drop it as well, once you get to it :-)

Ah, thanks for the heads-up - I've remove it, and instead check for the commands presence like the selinux test does.

bluca commented:
udev test is failing in semaphoreci, doesn't look related - I've got really bad luck with the CI today :-)

poettering commented on src/core/namespace.c:
```diff
@@ -1264,6 +1264,7 @@ int setup_namespace(
         _cleanup_(decrypted_image_unrefp) DecryptedImage *decrypted_image = NULL;
         _cleanup_(dissected_image_unrefp) DissectedImage *dissected_image = NULL;
         _cleanup_free_ void *root_hash = NULL;
+        _cleanup_free_ char *hash_device = NULL;
```
i must say I find "hash_device" a surprising name for this, given that this is most likely a plain file no? Maybe call this verity_data or verity_image?

bluca replied:
I named it after what cryptsetup/veritysetup calls it, for cross-tools consistency - from "man veritysetup":

> open <data_device> <name> **<hash_device>** <root_hash> create <name> <data_device> **<hash_device>** <root_hash>

But I do not mind one way or another, so I'll change to verity_data

poettering replied:
they have a different focus though than we I'd say. This stuff we do here is clearly intended for files, they focus more on block devices. I think we should stick to more generic terminology hence?

bluca replied:
Sure, no problem

poettering commented on src/dissect/dissect.c:
```diff
@@ -39,7 +41,8 @@ static void help(void) {
                "     --discard=MODE    Choose 'discard' mode (disabled, loop, all, crypto)\n"
-               "     --root-hash=HASH  Specify root hash for verity\n",
+               "     --root-hash=HASH  Specify root hash for verity\n"
+               "     --hash-device=PATH   Specify hash device for verity\n",
```
please rebreak the whole table, so that things are nicely aligned again

bluca replied:
sure, done

poettering commented on src/dissect/dissect.c:
```diff
@@ -127,6 +132,14 @@ static int parse_argv(int argc, char *argv[]) {
                 }
 
+                case ARG_HASH_DEVICE: {
+                        arg_hash_device = strdup(optarg);
```
leaks memory if arg_hash_device was already set. Please use: 

```c
if (free_and_strdup(&arg_hash_device, optarg) < 0)
        return log_oom();
```

bluca replied:
Ah I didn't think about the case where it's passed multiple times - fixed, thanks

poettering commented on src/dissect/dissect.c:
```diff
@@ -127,6 +132,14 @@ static int parse_argv(int argc, char *argv[]) {
                         break;
                 }
 
+                case ARG_HASH_DEVICE: {
```
The {} is unnecessary here as this block doesn't need any locally scoped variables

poettering commented on src/dissect/dissect.c:
```diff
@@ -225,7 +237,7 @@ static int run(int argc, char *argv[]) {
                         k = PARTITION_VERITY_OF(i);
                         if (k >= 0)
-                                printf(" %s verity", m->partitions[k].found ? "with" : "without");
+                                printf(" %s verity", m->partitions[k].found || m->verity ? "with" : "without");
```
given the test is a bit more complex now: could you please add a helper call dissected_image_has_verity() in dissect-image.c or so that returns a simple boolean  and internalizes the test? That would make things a lot more readable

bluca replied:
Sure, added 2 helpers - _has_verity and _can_do_verity

poettering commented on src/nspawn/nspawn.c:
```diff
@@ -1316,6 +1321,14 @@ static int parse_argv(int argc, char *argv[]) {
                 }
 
+                case ARG_HASH_DEVICE: {
+                        arg_hash_device = strdup(optarg);
```
as above, free_and_strdup()

poettering commented on src/nspawn/nspawn.c:
```diff
@@ -1316,6 +1321,14 @@ static int parse_argv(int argc, char *argv[]) {
                         break;
                 }
 
+                case ARG_HASH_DEVICE: {
```
{} unnecessary

poettering commented on src/nspawn/nspawn.c:
```diff
@@ -5129,12 +5142,11 @@ static int run(int argc, char *argv[]) {
+                                        arg_hash_device ? NULL : &arg_hash_device);
+                        if (r < 0) {
+                                return log_error_errno(r, "Failed to read verity artefacts for %s: %m", arg_image);
+                                goto finish;
```
the goto is never reached

bluca replied:
thanks, copypasta - fixed

poettering commented on test/units/testsuite-49.sh:
```diff
@@ -0,0 +1,22 @@
+set -ex
+set -o pipefail
+
+modprobe loop
```
this should be unnecessary

bluca replied:
I'll check again, might be leftover of putting the image together with all the extra modules

poettering commented on src/shared/dissect-image.c:
```diff
@@ -413,9 +413,11 @@ int dissect_image(
                         if (r < 0)
                                 return r;
 
+                        m->verity = !sd_id128_is_null(verity_uuid);
```
hmm, verity_uuid is set exactly when root_hash is non-NULL. I think it would be much nicer to test for that here directly:
`
```c
m->verity = !!root_hash;
```

After all for your case the verity UUID is not actually used. After all we only use it as method to find the right root fs with the right verity, and look for the uuid in the gpt part table. But if you have no gpt table then the uuid is not useful. So the root hash is actually hhat you really want to check.

bluca replied:
good point, fixed

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1223,17 +1226,17 @@ static int verity_partition(
 
-        if (!streq(v->fstype, "DM_verity_hash"))
+        if (!hash_device && !streq(v->fstype, "DM_verity_hash"))
                 return 0;
```
instead of duplicating the `!hash_device` check, maybe enclose both if checks in a surrounding `if (!hash_device) {…` block?

bluca replied:
Ok, fixed

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1409,56 +1414,101 @@ int decrypted_image_relinquish(DecryptedImage *d) {
 }
 
-int root_hash_load(const char *image, void **ret, size_t *ret_size) {
+int verity_metadata_load(const char *image, void **roothash, size_t *roothash_size, char **hashdevice) {
```
we generally prefix our return parameters with `ret_` to make clear that they are about output, not input. Hence please rename these `ret_root_hash`, `ret_root_hash_size` and `ret_verity_image` or so

bluca replied:
Ok, done

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1409,56 +1414,101 @@ int decrypted_image_relinquish(DecryptedImage *d) {
+                        if (e)
+                                n = e;
+                        else {
+                                e = endswith(fn, ".squashfs");
```
hmm, this looks awfully specific to the contents. So far we stuck to `.raw` which is a nicely generic suffix.

bluca replied:
Ok, will remove

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1409,56 +1414,101 @@ int decrypted_image_relinquish(DecryptedImage *d) {
+                        if (r == -ENOENT) {
+                                *roothash = NULL;
+                                *roothash_size = 0;
+                                return 0;
```
As per our coding style all functions should initialize all return parameters on sucess (and touch none of them on failure). In this case you only touch two of the three return parameters

bluca replied:
good point, I'll add a cleanup section at the end

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1409,56 +1414,101 @@ int decrypted_image_relinquish(DecryptedImage *d) {
+                        if (!IN_SET(r, -ENODATA, -EOPNOTSUPP, -ENOENT))
+                                return r;
+
+                        *hashdevice = (char *)malloc(strlen(image) + STRLEN(".hashdevice") + 1);
```
please do not clobber return params on failure. i.e. build the path in a separate variable, and only when you know everyting worked copy them into the the return parameters.

Also, to concat strings, use strjoin()

bluca replied:
I'll change to set the return only at the end. For the string, this is not really concatenating but overwriting the suffix, like it was already done for the roothash - doesn't look like strjoin would do the right thing?

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1409,56 +1414,101 @@ int decrypted_image_relinquish(DecryptedImage *d) {
+
+                        strcpy(n, ".hashdevice");
+
+                        r = access(*hashdevice, R_OK);
```
hmm, is this check really desirable? if the file is missing we can detect that when we want to use, why do this ahead of time?

smells like toctou waiting to happen

bluca replied:
Just following the same pattern for the root hash, where the file is read immediately. Will remove it.

poettering commented on test/units/testsuite-49.sh:
```diff
@@ -0,0 +1,22 @@
+
+cd /tmp
+
+mkdir -p img/usr/lib img/etc
```
we nowadays run most tests in the same image, hence each test should either use random test dirs or uniquely named ones, so that they are unlikely to conflict with other tests that might run in parallel. i.e. "img" is too generic a subdir, might conflict with another test.

bluca replied:
Isn't /tmp on a tmpfs? But anyway it's a good point, I'll create a random dir with mktemp, it seems good practice

bluca commented:
> Looks pretty good, but some comments.
> 
> Also needs man page additions for systemd-nspawn + systemd-mount

Thanks, all fixes pushed, including changes to nspawn's manpage which I had forgot about.
systemd-dissect doesn't have a manpage to update, though.

boucman commented:
> systemd-dissect doesn't have a manpage to update, though.

That's probably a left-over from when systemd-dissecct wasn't officially blessed and was still hidden in /usr/lib/systemd.

Please open a feature-request (or even better, a patch) so it's not forgotten

bluca commented:
> > systemd-dissect doesn't have a manpage to update, though.
> 
> That's probably a left-over from when systemd-dissecct wasn't officially blessed and was still hidden in /usr/lib/systemd.
> 
> Please open a feature-request (or even better, a patch) so it's not forgotten

I see - I think it's still hidden in /usr/lib/systemd though, so perhaps it's fine not to have it?

mrc0mmand commented:
CentOS 7 fail:

```
[    6.434143] systemd[1]: Starting systemd-update-utmp-runlevel.service...
[    6.434996] testsuite-49.sh[473]: + /usr/lib/systemd/systemd-dissect /tmp/tmp.o4smQN.raw
[    6.437858] systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1/unit/systemd_2dupdate_2dutmp_2drunlevel_2eservice interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=13 r>
[    6.438053] testsuite-49.sh[474]: + grep -q -F -f /usr/lib/os-release
...
[    6.457916] kernel: Dev loop0: unable to read RDB block 8
[    6.457927] kernel:  loop0: unable to read partition table
[    6.457933] kernel: loop0: partition table beyond EOD, truncated
[    6.457939] kernel: loop_reread_partitions: partition scan of loop0 () failed (rc=-5)
...
[    6.483263] testsuite-49.sh[440]: + mv /tmp/tmp.o4smQN.veritydata /tmp/tmp.o4smQN.foodata
...
[    6.490710] testsuite-49.sh[440]: + mv /tmp/tmp.o4smQN.roothash /tmp/tmp.o4smQN.foohash
[    6.493685] testsuite-49.sh[481]: + grep -q -F 'Found read-only '\''root'\'' partition of type squashfs with verity'
[    6.494651] testsuite-49.sh[482]: ++ cat /tmp/tmp.o4smQN.foohash
[    6.498420] testsuite-49.sh[480]: + /usr/lib/systemd/systemd-dissect /tmp/tmp.o4smQN.raw --root-hash=894ef88bdd131760327a7f9ce8aab738edc86ce85591c8771c574c06d7fa9ebc --verity-data=/tmp/tmp.o4smQN.foodata
[    6.507467] kernel: Dev loop0: unable to read RDB block 8
[    6.507479] kernel:  loop0: unable to read partition table
[    6.507487] kernel: loop0: partition table beyond EOD, truncated
[    6.507493] kernel: loop_reread_partitions: partition scan of loop0 () failed (rc=-5)
...
[    6.525672] testsuite-49.sh[486]: + grep -q -F -f /usr/lib/os-release
[    6.526424] testsuite-49.sh[487]: ++ cat /tmp/tmp.o4smQN.foohash
[    6.527091] testsuite-49.sh[485]: + /usr/lib/systemd/systemd-dissect /tmp/tmp.o4smQN.raw --root-hash=894ef88bdd131760327a7f9ce8aab738edc86ce85591c8771c574c06d7fa9ebc --verity-data=/tmp/tmp.o4smQN.foodata
[    6.533128] kernel: Dev loop0: unable to read RDB block 8
[    6.533139] kernel:  loop0: unable to read partition table
[    6.533145] kernel: loop0: partition table beyond EOD, truncated
[    6.533150] kernel: loop_reread_partitions: partition scan of loop0 () failed (rc=-5)
...
[    6.540283] testsuite-49.sh[485]: Failed to acquire image metadata: Protocol error
...
    6.553685] systemd[1]: Child 440 (bash) died (code=exited, status=1/FAILURE)
[    6.553787] systemd[1]: testsuite-49.service: Child 440 belongs to testsuite-49.service.
[    6.553892] systemd[1]: testsuite-49.service: Main process exited, code=exited, status=1/FAILURE
[    6.554029] systemd[1]: testsuite-49.service: Failed with result 'exit-code'.
```

Full journal: https://ci.centos.org/job/systemd-pr-build/12245/artifact//systemd-centos-ci/artifacts_EC3dFb/testsuite-logs-upstream.u0G/TEST-49-DISSECT/system.journal

bluca commented:
> Failed to acquire image metadata: Protocol error

Interesting error - the same image can be dissected fine on other OSes

bluca commented:
> > Failed to acquire image metadata: Protocol error
> 
> Interesting error - the same image can be dissected fine on other OSes

Wait, just happened once on my machine too, then worked when immediately re-run - what...?

```
$ sudo  buildx86/systemd-dissect --root-hash `cat img.squashfs.roothashfoo` --verity-data img.squashfs.veritydata img.squashfs
Found read-only 'root' partition of type squashfs with verity (/dev/block/7:55)
Failed to acquire image metadata: Protocol error
$ sudo  buildx86/systemd-dissect --root-hash `cat img.squashfs.roothashfoo` --verity-data img.squashfs.veritydata img.squashfs
Found read-only 'root' partition of type squashfs with verity (/dev/block/7:55)
OS Release: PRETTY_NAME=Debian GNU/Linux 10 (buster)
            NAME=Debian GNU/Linux
            VERSION_ID=10
            VERSION=10 (buster)
            VERSION_CODENAME=buster
            ID=debian
            HOME_URL=https://www.debian.org/
            SUPPORT_URL=https://www.debian.org/support
            BUG_REPORT_URL=https://bugs.debian.org/
```

bluca commented:
> > > Failed to acquire image metadata: Protocol error
> > 
> > 
> > Interesting error - the same image can be dissected fine on other OSes
> 
> Wait, just happened once on my machine too, then worked when immediately re-run - what...?
> 
> ```
> $ sudo  buildx86/systemd-dissect --root-hash `cat img.squashfs.roothashfoo` --verity-data img.squashfs.veritydata img.squashfs
> Found read-only 'root' partition of type squashfs with verity (/dev/block/7:55)
> Failed to acquire image metadata: Protocol error
> $ sudo  buildx86/systemd-dissect --root-hash `cat img.squashfs.roothashfoo` --verity-data img.squashfs.veritydata img.squashfs
> Found read-only 'root' partition of type squashfs with verity (/dev/block/7:55)
> OS Release: PRETTY_NAME=Debian GNU/Linux 10 (buster)
>             NAME=Debian GNU/Linux
>             VERSION_ID=10
>             VERSION=10 (buster)
>             VERSION_CODENAME=buster
>             ID=debian
>             HOME_URL=https://www.debian.org/
>             SUPPORT_URL=https://www.debian.org/support
>             BUG_REPORT_URL=https://bugs.debian.org/
> ```

With debug level, it looks like sometimes it fails to mount:

> Failed to mount /dev/block/7:55 (type squashfs) on /tmp/dissect-72co7c (MS_RDONLY|MS_NODEV ""): No such file or directory

bluca commented:
It seems to be a race with the creation of the symlink in /dev/block to the loop device for single-filesystems images. It happens without any of these changes too. For testing, added a sleep in dissected_image_mount and the issue goes away.
I'll see if I can figure out why /dev/block/foo is not created despite no error being raised.

bluca commented:
Unless I'm mistaken, it seems when ioctl(control, LOOP_CTL_GET_FREE) returns ( https://github.com/systemd/systemd/blob/master/src/shared/loop-util.c#L108 ) the corresponding /dev/block/x:y is not always there already. Adding some wait after that ioctl makes the problem go away. Does that sound familiar to anyone?

poettering commented:
> That's probably a left-over from when systemd-dissecct wasn't officially blessed and was still hidden in /usr/lib/systemd.
> 
> Please open a feature-request (or even better, a patch) so it's not forgotten

THere's a TODO list item to make it public. If you are up to it, move to to bindir, and add a man page.

poettering commented on man/systemd-nspawn.xml:
```diff
@@ -302,6 +302,9 @@
         hash partitions are set up if the root hash for them is specified using the <option>--root-hash=</option>
         option.</para>
 
+        <para>Single file system images can be opened using dm-verity if the integrity data is passed using the
```
maybe clarify with `(i.e. file systems without a surrounding partition table)` or so

bluca replied:
done

poettering commented on src/dissect/dissect.c:
```diff
@@ -61,6 +66,7 @@ static int parse_argv(int argc, char *argv[]) {
                 { "discard",   required_argument, NULL, ARG_DISCARD   },
                 { "root-hash", required_argument, NULL, ARG_ROOT_HASH },
                 { "fsck",      required_argument, NULL, ARG_FSCK      },
+                { "verity-data",   required_argument, NULL, ARG_VERITY_DATA   },
```
please rebreak table, so that everything is nicely aligned again

bluca replied:
missed this one, fixed

poettering commented on man/systemd-nspawn.xml:
```diff
@@ -390,8 +393,20 @@
+        <listitem><para>Takes the path to a data integrity (dm-verity) file. This option enables data integrity checks
+        using dm-verity, if a root-hash is passed and if the used image itself does not contains the integrity data.
+        The integrity data must be matched by the root hash. If this option is not specified, but a file with the
+        <filename>.veritydata</filename> suffix is found next to the image file, bearing otherwise the same name (except if
```
i think `.verity` would be a lot sexier

bluca replied:
For the argument too (--verity instead of --verity-data) or only the filename?

poettering replied:
only for the file suffix

bluca replied:
ack, done

poettering commented on src/nspawn/nspawn.c:
```diff
@@ -5129,12 +5140,11 @@ static int run(int argc, char *argv[]) {
+                        r = verity_metadata_load(arg_image, arg_root_hash ? NULL : &arg_root_hash, &arg_root_hash_size,
+                                        arg_verity_data ? NULL : &arg_verity_data);
+                        if (r < 0) {
+                                r = log_error_errno(r, "Failed to read verity artefacts for %s: %m", arg_image);
```
doesn't matter, but there's not much point in assigning the return value of the log call to `r`, given that the very same value is already in there...

bluca replied:
Removed - was copied from other usage in the same function

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1661,6 +1693,30 @@ int dissect_image_and_warn(
+        if (k >= 0 && image->partitions[k].found)
+                return true;
+        return false;
+}
```
actually, this isn't right. So previously "systemd-dissect" would show whether a *specific* partition has verity. This of course would only ever be true for the root partition, but it was supposed to be an inherent per-partition check. The `verity` and `can_verity` booleans in the DissectedImage structure indicated globally whether verity was available, i.e. whether verity is used at all, by any of the partitions.

With your change an image that consists of a GPT partition table with a root partition, a verity partition for it, pluse five others, would be shown as having verity for each one of these partitions, but that's simply incorrect: only the root partition actually has it.

I figure you should introduce a new, seperate field in DissectedImage called "external_verity" or so that encodes whether external verity is available...

and then only have:

```c
bool dissected_image_partition_has_verity(const DissectedImage *image, int p) {
        assert(image);

        if (p == PARTITION_ROOT && image->external_verity) /* Externally configured verity applies to the root partition only */
                return true;

        k = PARTITION_VERITY_OF(p);
        if (k < 0) /* This partition type has no verity concept */
                return false;

        return image->partitions[k].found;
}
```

Note that getting this right might not appear to matter much right now. However, I hope to add some special support for verity enabled /usr partitions (that are independent of the root partitions, so that one can combine a writable root with a verity protected /usr), and then this really starts to matter, since suddenly other partition types than just the root partition can have verity sensibly

bluca replied:
Interesting use case, hadn't thought of it, thanks for mentioning it.
Adding a new flag looks a bit awkward though - I'd like to propose another alternative first: the global has_verity is used only if the image has a single partition. Simple and self-contained, and should work with multiple verity partitions.
If you'd still prefer to have a separate flag instead, let me know and I can work on that instead.

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1409,56 +1416,81 @@ int decrypted_image_relinquish(DecryptedImage *d) {
+                        r = read_one_line_file(fn, &text);
+                        if (r == -ENOENT) {
+                                *ret_roothash = NULL;
+                                *ret_roothash_size = 0;
```
as before: initialize either all return params (on success) or none (on failure). Here you initialize them already when parts succeeded.

bluca replied:
I cheated a bit and moved the .verity block first, so that this snippet always comes last :-)

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1409,56 +1416,81 @@ int decrypted_image_relinquish(DecryptedImage *d) {
 
-        *ret = TAKE_PTR(k);
-        *ret_size = l;
+                verity_filename = (char *)malloc(strlen(image) + STRLEN(".veritydata") + 1);
```
new(char, …)

bluca replied:
Neat, updated, thanks

bluca commented:
@poettering so when dissecting a single-fs image, the mount source is set to the /dev/block/X:Y symlink rather than /dev/loopZ here: https://github.com/systemd/systemd/blob/master/src/shared/dissect-image.c#L422

As far as I can see, the problem is that this symlink is created asynchronously by udev here:
https://github.com/systemd/systemd/blob/master/src/udev/udev-node.c#L65

> Jun 03 17:44:12 bluca-lenovo systemd-udevd[38180]: loop0: Creating symlink '/dev/block/7:0' to '../loop0'

But this happens at some point after the image file descriptor has been associated with the loop device via ioctl. As far as I can see it's not guaranteed to be done by the time we then try to mount the loop device, hence it sometimes fails.

Any advice on what to do here? Can we use the loop device directly instead?

poettering commented:
I figure we need to do the loop_wait_for_partitions_to_appear() dance also for no-partition images, it currently is explicitly excluded.

bluca commented:
> I figure we need to do the loop_wait_for_partitions_to_appear() dance also for no-partition images, it currently is explicitly excluded.

Ah, thanks - it was right if front of my face and yet missed that call. Added a commit to do the check unconditionally - although waiting for udev sounds expensive, and I might come back to this in the future and look for a shortcut if it turns out to be measurably more expensive.

bluca commented:
> hmmm, i wonder if we should explicitly refuse the case where we find a GPT partition table but an explicit verity data file is specified. Because it's not clear what people are OK if the other partitions have no verity protection or not, and it sounds like a weird combo that could potentially be a vulnerability.
> 
> Maybe refuse that case for now. If there's actaully a usecase for it, and the security implications are fully understood we can revisit this of course, but for now I'd be rather safe than sorry and say: either you have GPT and then root and verity partition in there. Or you have no partition table and specify both directly. But having GPT with root included and verity externally should be refused.

Ok, I've added back the access() check on the verity file for this, and a new flag for the dissect_image function that instructs to accept only single-filesystem images.

bluca commented:
Both CI failures look unrelatated:

> +./fuzzit create job --type regression --branch PR-16046 --revision d46f9ee054e83ab22d0dced7eaaf5ff0b03a8324 -e ASAN_OPTIONS=redzone=32:print_summary=1:handle_sigill=1:allocator_release_to_os_interval_ms=500:print_suppressions=0:strict_memcmp=1:allow_user_segv_handler=0:allocator_may_return_null=1:use_sigaltstack=1:handle_sigfpe=1:handle_sigbus=1:detect_stack_use_after_return=1:alloc_dealloc_mismatch=0:detect_leaks=1:print_scariness=1:max_uar_stack_size_log=16:handle_abort=1:check_malloc_usable_size=0:quarantine_size_mb=64:detect_odr_violation=0:handle_segv=1:fast_unwind_on_fatal=0 -e UBSAN_OPTIONS=print_stacktrace=1:print_summary=1:halt_on_error=1:silence_unsigned_overflow=1 fuzz-journald-stream-asan-ubsan out/fuzz-journald-stream ./out/src/shared/libsystemd-shared-245.so
2020/06/03 23:56:27 Get https://app.fuzzit.dev/createCustomToken?api_key=af6992074353998676713818cc6435ef4a750439932dab58b51e9354d6742c54d740a3cd9fc1fc001db82f51734a24bc: dial tcp: lookup app.fuzzit.dev on 169.254.169.254:53: server misbehaving


> autopkgtest-virt-ssh: WARNING: ssh connection failed. Retrying in 3 seconds...

bluca commented:
Rebased to fix merge conflict from another PR

poettering commented on src/shared/dissect-image.c:
```diff
@@ -432,7 +435,8 @@ int dissect_image(
 
                         return 0;
-                }
+                } else if (flags & DISSECT_IMAGE_NO_GPT)
```
this cannot happen, the first branch already covers this

bluca replied:
Removed

poettering commented on src/shared/dissect-image.h:
```diff
@@ -63,6 +63,7 @@ typedef enum DissectImageFlags {
         DISSECT_IMAGE_NO_UDEV             = 1 << 9,  /* Don't wait for udev initializing things */
         DISSECT_IMAGE_RELAX_VAR_CHECK     = 1 << 10, /* Don't insist that the UUID of /var is hashed from /etc/machine-id */
         DISSECT_IMAGE_FSCK                = 1 << 11, /* File system check the partition before mounting (no effect when combined with DISSECT_IMAGE_READ_ONLY) */
+        DISSECT_IMAGE_NO_GPT              = 1 << 12, /* Only recognize single file system images */
```
i think we should name this more precisely, people might think this would just turn off gpt and leave mbr partition table support on. i.e. maybe call this DISSECT_IMAGE_NO_PARTITION_TABLE? or DISSECT_WITHOUT_PARTITION_TABLE or so?

bluca replied:
Makes sense, done

bluca replied:
Makes sense, done

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1409,56 +1417,89 @@ int decrypted_image_relinquish(DecryptedImage *d) {
+                        if (errno != ENOENT)
+                                return -errno;
+                        free(verity_filename);
+                        verity_filename = NULL;
```
`verity_filename = mfree(verity_filename);`

bluca replied:
Thanks, fixed - hadn't spotted that helper before, neat

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1409,56 +1417,89 @@ int decrypted_image_relinquish(DecryptedImage *d) {
+                *ret_roothash_size = roothash_decoded_size;
+        }
+        if (verity_filename)
+                *ret_verity_data = TAKE_PTR(verity_filename);
```
as mentioned: our rule is to unconditionally initialize returm values on success. 

this, these if checks should just not be done, but the three asignments should be done here unconditionally. (you can then drop the early reset to NULL/0 further up too) The code becomes a lot simpler, and the return params are initialized at once place only.

bluca replied:
Ok - I've removed the earlier reset, and changed the check to look at the return buffers instead.

It is possible for one return buffer to be passed and one to be null, as it's legal for example to call nspawn --root-hash (thus this function gets a null as we don't want to overwrite what the user set explicitly) and let the auto-discovery find the verity file, or viceversa

poettering commented on src/shared/dissect-image.c:
```diff
@@ -413,9 +414,11 @@ int dissect_image(
                         if (r < 0)
                                 return r;
 
+                        m->verity = !!root_hash;
```
so far, the "verity" boolean so far was set when both a root hash hash was set and verity data was found in the image. And "can_verity" was set if verity data was found but no root hash specified.

I figure dissect_image() should also have a verity_data argument that it uses for initializing can_verity.

i.e.

```c
m->verity = verity_data && root_hash;
m->can_verity = !!verity_data;
```

bluca replied:
Ok, done

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1661,6 +1698,40 @@ int dissect_image_and_warn(
+                /* image->can_verity is meaningful on a per-partition check only with a single file system */
+                for (unsigned i = PARTITION_ROOT; i < _PARTITION_DESIGNATOR_MAX; i++)
+                        if (image->partitions[i].found)
+                                return false;
```
this loop will always return false, since if there's no part table we detect the whole image as `PARTITION_ROOT` designator

bluca replied:
Forgot the ```+ 1``` in the initialization - fixed

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1661,6 +1701,40 @@ int dissect_image_and_warn(
         }
 }
 
+bool dissected_image_can_do_verity(const DissectedImage *image, unsigned partition_number) {
```
"partition_number" is a misnomer. It's a designator type, not a partition index after all

bluca replied:
changed to partition_deisgnator

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1661,6 +1701,40 @@ int dissect_image_and_warn(
+                /* image->can_verity is meaningful on a per-partition check only with a single file system */
+                for (unsigned i = PARTITION_ROOT + 1; i < _PARTITION_DESIGNATOR_MAX; i++)
+                        if (image->partitions[i].found)
+                                return false;
```
Hmm, I think it would be prettier to just store a boolean flag `naked_file_system` or so in the `DissectImage` structure that indicates whether we have a partition table or not, instead of checking this after the fact. After all this code could not distuingish the case "gpt with only one root partition" from "no partition table". This check then becomes a lot easier:

```
if (image->naked_file_system)
        return designator == PARTITION_ROOT && image->can_verity;

return PARTITION_VERITY_OF(designator) >= 0;
```

bluca replied:
Ok, no strong preference - done

poettering commented on src/shared/dissect-image.c:
```diff
@@ -1661,6 +1701,40 @@ int dissect_image_and_warn(
+                        if (image->partitions[i].found)
+                                return false;
+                return true;
+        }
```
with the `naked_file_system` boolean suggested above this would become:

```c
if (image->naked_file_system)
        return designator == PARTITION_ROOT && image->verity;

k = PARTITION_VERITY_OF(designator);
return k >= 0 && image->partitions[k].found;
```

bluca replied:
Done

bluca commented on src/nspawn/nspawn.c:
```diff
@@ -1316,6 +1321,12 @@ static int parse_argv(int argc, char *argv[]) {
                 }
 
+                case ARG_VERITY_DATA:
+                        if (free_and_strdup(&arg_verity_data, optarg) < 0)
```
One more useful helper I missed. There's really a lot of em :-) Fixed

bluca commented on src/shared/dissect-image.c:
```diff
@@ -391,8 +392,9 @@ int dissect_image(
 
-        if (!(flags & DISSECT_IMAGE_GPT_ONLY) &&
-            (flags & DISSECT_IMAGE_REQUIRE_ROOT)) {
+        if ((!(flags & DISSECT_IMAGE_GPT_ONLY) &&
```
Makes sense, done

bluca commented on src/shared/dissect-image.c:
```diff
@@ -1409,56 +1420,85 @@ int decrypted_image_relinquish(DecryptedImage *d) {
-                        *ret = NULL;
-                        *ret_size = 0;
-                        return 0;
+                r = access(verity_filename, R_OK);
```
done

poettering commented on src/dissect/dissect.c:
```diff
@@ -127,6 +134,12 @@ static int parse_argv(int argc, char *argv[]) {
                 }
 
+                case ARG_VERITY_DATA:
+                        r = parse_path_argument_and_warn(optarg, true, &arg_verity_data);
```
i don't think the second argument should be true.

It should only be used when parsing chroot paths. But here this doesn't apply

bluca replied:
Ok, I misunderstood what the flag was for, fixed

poettering commented on src/nspawn/nspawn.c:
```diff
@@ -1316,6 +1321,12 @@ static int parse_argv(int argc, char *argv[]) {
                 }
 
+                case ARG_VERITY_DATA:
+                        r = parse_path_argument_and_warn(optarg, true, &arg_verity_data);
```
as above

bluca replied:
fixed

poettering commented:
excellent! thanks!

bluca commented:
CentOS/Arch failure is in TEST-02-CRYPTSETUP, does not seem to be related
